### PR TITLE
Simplify implementation of ad_hoc dataset

### PR DESCRIPTION
**Pull request by andrenmelo**
### Summary
This PR simplifies the implementation of `ad_hoc_dataset`. This rewrite also makes it easier to generalise to higher dimensions, although before doing that I'd like to discuss how to go about it with y'all.

A couple of remarks:
- I've checked that this code reproduces the figures in the [tutorial](https://qiskit.org/documentation/machine-learning/tutorials/03_quantum_kernel.html).
- I made `sample_ad_hoc_data` a private function, which I think makes more sense. AFAICT this function is not used anywhere anyhow.

### CODE REVIEW

File: qiskit_machine_learning/datasets/ad_hoc.py [hunk 1]
```diff
@@ -329,40 +129,37 @@ def ad_hoc_data(
...
+
+def _plot_ad_hoc_data(x_total, y_total, training_size):
+    try:
+        import matplotlib.pyplot as plt
+    except ImportError as ex:
+        raise MissingOptionalLibraryError(
+            libname="Matplotlib",
+            name="ad_hoc_data",
+            pip_install="pip install matplotlib",
+        ) from ex
```
Comment (adekusar-drl) [inline qiskit_machine_learning/datasets/ad_hoc.py:156]: Please take a look at the recent changes in terra. You can simplify the function by adding `@HAS_MATPLOTLIB.require_in_call` to the the function definition and remove `try ... except...` in the function.

The function should look like:
```python
from qiskit.utils.optionals import HAS_MATPLOTLIB

@HAS_MATPLOTLIB.require_in_call
def _plot_ad_hoc_data(x_total, y_total, training_size):
    # your plotting code
```
And bump terra in the `requirements.txt` to `0.20`.

**Review by adekusar-drl — COMMENTED** [review]
Thanks a lot and sorry for the delay. I don't have many comments here:
- Can you please make use of `HAS_MATPLOTLIB`?
- If you don't mind, please, add some text comments in the code that describe how the dataset is generated since the code is not easy to follow. This is helpful for code readers.
- If you have any ideas how to improve the unit tests, would be nice to hear. But this can be done separately.

**Comment by andrenmelo** [comment]
@adekusar-drl Implemented some of your suggestions!

Regarding additional unit tests: one thing I can think of is to freeze the dataset used for the [kernel paper](https://arxiv.org/pdf/1804.11326.pdf) (also used in the [kernel tutorial](https://qiskit.org/documentation/machine-learning/tutorials/03_quantum_kernel.html)) and then check whether the code reproduces it. How does that sound?

**Comment by adekusar-drl** [comment]
@andrenmelo how is it going? do you need any help with this PR?

**Comment by andrenmelo** [comment]
Hi @adekusar-drl, thanks for checking in! I was busy with moving so I haven't made any progress. I was just about to restart work on the PR though, so I will probably push something later today.

**Comment by andrenmelo** [comment]
@adekusar-drl Should be good to go. Should I write a Reno note as well? If so, which category would be appropriate?

**Comment by adekusar-drl** [comment]
@andrenmelo Thanks for fixing issues in the PR. I don't think there's a need for reno. You changed the way how adhoc is generated, keeping the existing features. I think the PR is in a good shape. I will take a look later on today/tomorrow.

File: qiskit_machine_learning/datasets/ad_hoc.py [hunk 2]
```diff
@@ -319,40 +193,31 @@ def ad_hoc_data(
...
-    sample_train = np.reshape(sample_train, (2 * test_size, n))
-    test_input = {key: (sample_train[label_train == k, :])[:] for k, key in enumerate(class_labels)}
-    return test_input
+def _sample_ad_hoc_data(sample_total, xvals, num_samples, n):
+    count = sample_total.shape[0]
+    sample_a, sample_b = [], []
+    for i, sample_list in enumerate([sample_a, sample_b]):
+        label = 1 if i == 0 else -1
+        while len(sample_list) < num_samples:
+            draws = tuple(algorithm_globals.random.choice(count) for i in range(n))
```
Comment (1ucian0) [inline qiskit_machine_learning/datasets/ad_hoc.py:202]: Not for this PR, is there any plan to have a way to control the seed in this undeterministic parts?

**Review by adekusar-drl — APPROVED** [review]

**Comment by adekusar-drl** [comment]
@andrenmelo Thanks lot, looks good!

**Pull request merged**